docs/reinterpreting_effects.md

```diff
@@ -0,0 +1,362 @@
+
+## "Stacking" effects
+
+### The production use-case
+
+Now that we have these 2 mini-DSL effect types established, we need to stitch them together.
+
+Let's take a moment to think about what could go wrong with an HTTP API from which we plan to fetch some JSON and convert it into a list of `CatFact`s.
+
+We can conceive that the server might occasionally return a malformed JSON response:
+
+``` haskell
+data JsonParseError = JsonParseError String
+  deriving (Show, Eq)
+
+decodeOrThrow :: (Has (Throw JsonParseError) sig m, FromJSON a) => L.ByteString -> m a
+decodeOrThrow = either (throwError . JsonParseError) pure . eitherDecode
+```
+
+A more HTTP-centric issue is that we might receive a content type we can't use. In this case, anything that's not `application/json`:
+
+``` haskell
+data InvalidContentType = InvalidContentType String
+  deriving (Show, Eq)
+
+```
+
+Now we need to support fetching JSON given an HTTP request. We have no guarantee that an arbitrary HTTP request
+will actually return JSON, so for this implementation we have to account for failure conditions. This provides
+a great opportunity to see how effect handlers can actually rely on *multiple underlying effects*!
+
+``` haskell
+runCatFactsApi :: CatFactsApiC m a -> m a
+runCatFactsApi = runCatFactsApiC
+
+newtype CatFactsApiC m a = CatFactsApiC { runCatFactsApiC :: m a }
+ deriving newtype
+      ( Monad
+      , Functor
+      , Applicative
+      , MonadIO
+      , Alternative
+      )
+
+catFactsEndpoint :: HTTP.Request
+catFactsEndpoint = HTTP.parseRequest_ "https://cat-fact.herokuapp.com/facts/random"
+
+instance ( Has (Http :+: Throw JsonParseError :+: Throw InvalidContentType) sig m
+         , Algebra sig m
+         ) =>
+         Algebra (CatFactClient :+: sig) (CatFactsApiC m) where
+  eff (L (ListFacts numberOfFacts k)) = do
+    resp <- sendRequest (catFactsEndpoint { HTTP.queryString = "?amount=" <> B.pack (show numberOfFacts) })
+    case lookup hContentType (HTTP.responseHeaders resp) of
+      Just "application/json; charset=utf-8" -> decodeOrThrow (HTTP.responseBody resp) >>= k
+      other -> throwError (InvalidContentType (show other))
+  eff (R other) = CatFactsApiC (eff (handleCoercible other))
+```
+
+We implement a `CatFacts` effect handler that depends on _three_ underlying effects:
+
+1. `Http` - we need to be able to make requests
+2. `Throw JsonParseError` - we need to be able to signal that some aspect of the JSON wasn't what we expected.
+3. `Throw InvalidContentType` - we need to be able to signal what we got wasn't JSON at all!
+
+The nice aspect of this is that we have neatly contained the failure scenarios to their relevant strata rather than leaking them into the higher-level abstraction (`listFacts`)!
+
+Now we need to support performing HTTP requests:
+
+``` haskell
+runHttp :: HttpC m a -> m a
+runHttp = runHttpC
```

Comment:
We’ve taken to avoiding `C` suffixes if possible, so I’d recommend axing this and changing the newtype’s record selector to `runHttp`.